### PR TITLE
Separate attribute and method lookups

### PR DESCRIPTION
**Pull request by samscott89**
Attribute lookups and method calls (with no args) are no longer elided into the same thing.
`x.foo` will only look for an attribute on `x` called `foo`, and `x.foo()` will only look for a method on `x` called `foo`.

Todo:
- [x] Rust changes
- [x] Python
- [x] Ruby
- [x] Java
- [x] Javascript
- [x] Test for returning method as attribute
- [x] Documentation
- [x] Examples/doctests

PR checklist:

- [x] Added changelog entry.

### CODE REVIEW

File: languages/java/oso/src/main/java/com/osohq/oso/Query.java [hunk 1]
```diff
@@ -192,24 +199,27 @@ public void registerCall(String attrName, List<Object> args, long callId, JSONOb
...
+                    throw new Exceptions.InvalidCallError(cls.getName(), attrName, argTypes);
+                }
+            } else {
                 // Look for a field with the given name.
                 try {
                     Field field = cls.getField(attrName);
                     result = field.get(instance);
                 } catch (NoSuchFieldException f) {
-                    throw new Exceptions.InvalidCallError(cls.getName(), attrName, argTypes);
+                    throw new Exceptions.InvalidCallError(cls.getName(), attrName);
```
Comment (gj) [inline languages/java/oso/src/main/java/com/osohq/oso/Query.java:222]: Feels like this should be a distinct type of error, something like `Exceptions.InvalidFieldAccessError`

File: polar-core/src/vm.rs [hunk 2]
```diff
@@ -988,12 +995,18 @@ impl PolarVirtualMachine {
...
-        let (field_name, args) = match &field.value() {
+        let (field_name, args) = match self.deref(field).value() {
             Value::Call(Predicate { name, args }) => (
                 name.clone(),
-                args.iter().map(|arg| self.deref(arg)).collect(),
+                Some(args.iter().map(|arg| self.deref(arg)).collect()),
             ),
-            _ => unreachable!("call must be a predicate"),
+            Value::String(field) => (Symbol(field.clone()), None),
+            v => {
```
Comment (gj) [inline polar-core/src/vm.rs:1004]: Is this path actually reachable?

File: polar-core/src/vm.rs [hunk 3]
```diff
@@ -1374,15 +1387,15 @@ impl PolarVirtualMachine {
...
         assert_eq!(args.len(), 3);
         let object = self.deref(&args[0]);
-        let field = &args[1];
+        let field = self.deref(&args[1]);
         let value = &args[2];
 
         match object.value() {
             // Push a `Lookup` goal for simple field lookups on dictionaries.
-            Value::Dictionary(dict) if !matches!(field.value(), Value::Call(predicate) if !predicate.args.is_empty()) => {
+            Value::Dictionary(dict) if matches!(field.value(), Value::String(_) | Value::Variable(_)) => {
```
Comment (gj) [inline polar-core/src/vm.rs:1395]: We allow the RHS of a lookup to be a variable at this point?

File: languages/java/oso/src/main/java/com/osohq/oso/Query.java (hunk 1 above)
Comment (samscott89) [reply languages/java/oso/src/main/java/com/osohq/oso/Query.java:222]: Agreed, just did the same in TS.

File: polar-core/src/vm.rs (hunk 2 above)
Comment (samscott89) [reply polar-core/src/vm.rs:1004]: Yep, I changed the parser so a `x.y` parses as `.(x, "y")`

Comment (gj) [reply polar-core/src/vm.rs:1004]: To be clear, I'm asking about the catch-all path here, `v => {`

File: polar-core/src/vm.rs (hunk 3 above)
Comment (gj) [reply polar-core/src/vm.rs:1395]: That's really cool

File: polar-core/src/vm.rs (hunk 2 above)
Comment (samscott89) [reply polar-core/src/vm.rs:1004]: Ohh right. I guess you could if you do `x.(1)` or something, or manually construct `.(d, 1, value)`.

File: languages/python/oso/polar/query.py [hunk 4]
```diff
@@ -99,9 +94,15 @@ def handle_external_call(self, data):
...
-            if callable(attr):  # If it's a function, call it with the args.
+            if (
+                callable(attr) and not data["args"] is None
+            ):  # If it's a function, call it with the args.
+                args = [self.host.to_python(arg) for arg in data["args"]]
                 result = attr(*args)
+            elif not data["args"] is None:
+                raise RuntimeError(
+                    f"tried to call '{attribute}' but it is not callable"
+                )
```
Comment (gj) [inline languages/python/oso/polar/query.py:105]: nit: would prefer this as a new exception type that can be caught by users of the library

File: languages/java/oso/src/main/java/com/osohq/oso/Exceptions.java [hunk 5]
```diff
@@ -194,6 +194,16 @@ public InvalidCallError(String msg) {
...
     }
 
+    public static class InvalidAttributeError extends PolarRuntimeException {
+        public InvalidAttributeError(String className, String attrName) {
+            super("Invalid attribute `" + attrName + "` on class " + className);
+        }
+
+        public InvalidAttributeError(String msg) {
+            super(msg);
+        }
```
[the remark this answers is not in the document]
Comment (samscott89) [reply languages/java/oso/src/main/java/com/osohq/oso/Exceptions.java:204]: We don't you're right. I just copied it from the InvalidCallError

File: languages/js/src/Polar.test.ts [hunk 6]
```diff
@@ -200,6 +209,7 @@ describe('#registerClass', () => {
         'animal',
       ]);
       expect(qvar(p, `what_is(${canine}, r)`, 'r')).toStrictEqual([
+        undefined, // Canine has no species, so looking up the 'species' prop returns undefined.
```
[the remark this answers is not in the document]
Comment (samscott89) [reply languages/js/src/Polar.test.ts:212]: I guess this is where we're inconsistent on handling None/nil/undefined?

[the remark this answers is not in the document]
Comment (gj) [reply languages/js/src/Polar.test.ts:212]: Not sure what the broader implications are, but I just added this comment in the code (and highlighted it here) to raise awareness of that behavior.

File: languages/java/oso/src/main/java/com/osohq/oso/Exceptions.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (samscott89) [reply languages/java/oso/src/main/java/com/osohq/oso/Exceptions.java:204]: It's not doing much harm. Leaving it for now.

File: languages/js/src/Polar.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (samscott89) [reply languages/js/src/Polar.test.ts:212]: Thanks. I linked it to the Notion task we have for handling None better

File: languages/python/oso/polar/query.py (hunk 4 above)
Comment (samscott89) [reply languages/python/oso/polar/query.py:105]: This is consistent with how other exceptions are done in the Python library. Don't want to solve here.

**Pull request merged**